### PR TITLE
Linkage Checker exclusion file XML structure validation using Relax NG

### LINKED ISSUE #1277: exclusion file: Relax NG definition
exclusion file: Relax NG definition

Schema validation and conversion based on RELAX NG: https://github.com/relaxng/jing-trang

org.relaxng:jing:20181222

# Jing

## command line tool

RELAX NG tutorial: https://relaxng.org/tutorial-20011203.html

```
$ sudo apt-get install jing
...
suztomo@suxtomo24:~/beam6/hello-jing$ jing rng.xml addressbook.xml
[warning] /usr/bin/jing: Unable to locate avalon-framework in /usr/share/java
[warning] /usr/bin/jing: Unable to locate batik-all in /usr/share/java
/usr/local/google/home/suztomo/beam6/hello-jing/addressbook.xml:6:9: error: element "tomo" not allowed anywhere; expected the element end-tag or element "card"
suztomo@suxtomo24:~/beam6/hello-jing$ cat addressbook.xml 
<addressBook>
  <card>
    <name>John Smith</name>
    <email>js@example.com</email>
  </card>
  <tomo>foo</tomo>
  <card>
    <name>Fred Bloggs</name>
    <email>fb@example.net</email>
  </card>
</addressBook>
suztomo@suxtomo24:~/beam6/hello-jing$ cat rng.xml 
<element name="addressBook" xmlns="http://relaxng.org/ns/structure/1.0">
  <zeroOrMore>
    <element name="card">
      <element name="name">
        <text/>
      </element>
      <element name="email">
        <text/>
      </element>
    </element>
  </zeroOrMore>
</element>

```

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1277

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 1]
```diff
@@ -17,8 +17,14 @@
 package com.google.cloud.tools.opensource.classpath;
 
 import com.google.common.collect.ImmutableList;
+import com.thaiopensource.util.SinglePropertyMap;
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: The package name of this library is strange. This library org.relaxng:jing is from https://relaxng.org/jclark/api/jing/index.html

> You can use Jing's native API.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 2]
```diff
@@ -62,11 +68,25 @@
   static ImmutableList<LinkageErrorMatcher> parse(Path exclusionFile)
       throws SAXException, IOException {
 
+    // DraconianErrorHandler throws SAXException upon invalid structure
+    ValidationDriver validationDriver =
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: I'm really not convinced that we want to do this when processing the documents.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: Memo:

Cost
- performance for high qps services. (not this case)
- too restrictive. For example, users cannot put their exclusion files in pom.xml.
- adding dependencies. (fine)
- adding validation makes a system more fragile

File: dependencies/pom.xml [hunk 3]
```diff
@@ -79,6 +79,11 @@
       <artifactId>httpclient</artifactId>
       <version>4.5.12</version>
     </dependency>
+    <dependency>
+      <groupId>org.relaxng</groupId>
```
Comment (elharo) [inline dependencies/pom.xml]: Can you find a license for this? I don't see one anywhere. I think it's BSD but I don't see a statement about that.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 4]
```diff
@@ -62,11 +68,25 @@
   static ImmutableList<LinkageErrorMatcher> parse(Path exclusionFile)
       throws SAXException, IOException {
 
+    // DraconianErrorHandler throws SAXException upon invalid structure
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: You might want to extract all this into a helper method, just for clarity.

File: dependencies/src/main/resources/linkage-checker-exclusion-relax-ng.xml [hunk 5]
```diff
@@ -0,0 +1,56 @@
+<element name="LinkageCheckerFilter" xmlns="http://relaxng.org/ns/structure/1.0">
```
Comment (elharo) [inline dependencies/src/main/resources/linkage-checker-exclusion-relax-ng.xml]: copyright

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParserTest.java [hunk 6]
```diff
@@ -192,15 +192,44 @@ public void testParse_sourceAndTarget_unmatch()
...
 
   @Test
-  public void testParse_namespaceException() throws URISyntaxException, IOException, SAXException {
+  public void testParse_namespaceException() throws URISyntaxException, SAXException, IOException {
     Path exclusionFile = absolutePathOfResource("exclusion-sample-rules/target-with-namespace.xml");
     try {
       ImmutableList<LinkageErrorMatcher> matchers = ExclusionFileParser.parse(exclusionFile);
       fail();
-    } catch (SAXException ex) {
+    } catch (SAXParseException ex) {
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParserTest.java]: In this pattern, we sometimes name the exception `expected`

File: dependencies/src/main/resources/linkage-checker-exclusion-relax-ng.xml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                </attribute>
+                <attribute name="className">
+                  <text/>
+                </attribute>
+              </element>
+            </choice>
+          </element>
+        </optional>
+        <optional>
+        <element name="Source">
```
Comment (elharo) [inline dependencies/src/main/resources/linkage-checker-exclusion-relax-ng.xml]: indent +2

File: dependencies/src/main/resources/linkage-checker-exclusion-relax-ng.xml [hunk 8]
```diff
@@ -0,0 +1,71 @@
+<!--
+  ~ Copyright 2020 Google LLC.
```
Comment (suztomo) [inline dependencies/src/main/resources/linkage-checker-exclusion-relax-ng.xml]: Added copyright

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 9]
```diff
@@ -62,12 +68,29 @@
   static ImmutableList<LinkageErrorMatcher> parse(Path exclusionFile)
       throws SAXException, IOException {
 
+    File exclusion = exclusionFile.toFile();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: why convert to a File? Can we stick to Paths all the way through?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 10]
```diff
@@ -62,12 +68,29 @@
...
+    validate(exclusion);
+
     XMLReader xmlReader = XMLReaderFactory.createXMLReader();
     ExclusionFileHandler handler = new ExclusionFileHandler();
     xmlReader.setContentHandler(handler);
 
-    InputSource inputSource = new InputSource(new FileInputStream(exclusionFile.toFile()));
+    InputSource inputSource = new InputSource(new FileInputStream(exclusion));
     xmlReader.parse(inputSource);
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: If you call `inputSource.setSystemId(path.toURI())` or some such thing here, error reporting will improve.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 11]
```diff
@@ -62,12 +68,29 @@
...
 
+    File exclusion = exclusionFile.toFile();
+    validate(exclusion);
+
     XMLReader xmlReader = XMLReaderFactory.createXMLReader();
     ExclusionFileHandler handler = new ExclusionFileHandler();
     xmlReader.setContentHandler(handler);
 
-    InputSource inputSource = new InputSource(new FileInputStream(exclusionFile.toFile()));
+    InputSource inputSource = new InputSource(new FileInputStream(exclusion));
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: Files.newInputStream(path)

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java [hunk 12]
```diff
@@ -62,12 +68,29 @@
...
     xmlReader.parse(inputSource);
     return handler.getMatchers();
   }
+
+  private static void validate(File file) throws IOException, SAXException {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: This reads the file twice: once to validate and once to parse. Can this be combined into a single pass?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java (hunk 9 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: Updated to use Files.newInputStream.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java (hunk 11 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java (hunk 10 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java (hunk 12 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileParser.java]: I didn't think of a way to do that. ValidationDriver does not give a method related to XMLReader.

I thought reusing InputSource, but it's not intended to be reused:

>  so applications should not attempt to re-use such streams after they have been handed to a parser.

https://docs.oracle.com/javase/7/docs/api/org/xml/sax/InputSource.html#setPublicId(java.lang.String)

File: dependencies/pom.xml [hunk 13]
```diff
@@ -79,6 +79,16 @@
...
+      <artifactId>jing</artifactId>
+      <version>20181222</version>
+    </dependency>
+    <dependency>
+      <groupId>isorelax</groupId>
+      <artifactId>isorelax</artifactId>
```
Comment (suztomo) [inline dependencies/pom.xml]: http://iso-relax.sourceforge.net/ :  the MIT license

**Comment by suztomo** [comment]
@elharo PTAL. Isorelax worked good. Thank you for info.

**Pull request merged**